test-data/unit/check-overloading.test

```diff
@@ -5339,3 +5339,191 @@ def register(cls: Any) -> Any: return None
+    def f2(g: str) -> C: ...
+def f2(g: Any) -> Any: ...
+reveal_type(f2(42))  # N: Revealed type is "__main__.A"
+reveal_type(f2("Hello"))  # N: Revealed type is "__main__.C"
+
+@overload
+def f3(g: int) -> A: ...
+@overload
+def f3(g: str) -> B: ...
+if True:
+    def f3(g: Any) -> Any: ...
+reveal_type(f3(42))  # N: Revealed type is "__main__.A"
+reveal_type(f3("Hello"))  # N: Revealed type is "__main__.B"
+
+if True:
+    @overload
+    def f4(g: int) -> A: ...
+@overload
+def f4(g: str) -> B: ...
+def f4(g: Any) -> Any: ...
+reveal_type(f4(42))  # N: Revealed type is "__main__.A"
+reveal_type(f4("Hello"))  # N: Revealed type is "__main__.B"
+
+if True:
+    # Some comment
+    @overload
+    def f5(g: int) -> A: ...
+    @overload
+    def f5(g: str) -> B: ...
+def f5(g: Any) -> Any: ...
+reveal_type(f5(42))  # N: Revealed type is "__main__.A"
+reveal_type(f5("Hello"))  # N: Revealed type is "__main__.B"
+
+[case testOverloadIfNotMatching]
+from typing import overload, Any
+
+class A: ...
+class B: ...
+class C: ...
+
+@overload  # E: An overloaded function outside a stub file must have an implementation
+def f1(g: int) -> A: ...
+@overload
+def f1(g: bytes) -> B: ...
+if True:
+    @overload  # E: Name "f1" already defined on line 7 \
+               # E: Single overload definition, multiple required
+    def f1(g: str) -> C: ...
+    pass  # Some other action
+def f1(g: Any) -> Any: ...  # E: Name "f1" already defined on line 7
+reveal_type(f1(42))  # N: Revealed type is "__main__.A"
+reveal_type(f1("Hello"))  # E: No overload variant of "f1" matches argument type "str" \
+                          # N: Possible overload variants: \
+                          # N:     def f1(g: int) -> A \
+                          # N:     def f1(g: bytes) -> B \
+                          # N: Revealed type is "Any"
+
+if True:
+    pass  # Some other action
+    @overload  # E: Single overload definition, multiple required
+    def f2(g: int) -> A: ...
+@overload  # E: Name "f2" already defined on line 21
+def f2(g: bytes) -> B: ...
+@overload
+def f2(g: str) -> C: ...
+def f2(g: Any) -> Any: ...
+reveal_type(f2(42))  # N: Revealed type is "__main__.A"
+reveal_type(f2("Hello"))  # N: Revealed type is "__main__.A" \
+                          # E: Argument 1 to "f2" has incompatible type "str"; expected "int"
+
+[case testOverloadIfOldStyle]
+# flags: --always-false var_false --always-true var_true
+from typing import overload, Any
+
+class A: ...
+class B: ...
+
+var_true = True
+var_false = False
+
+if var_false:
+    @overload
+    def f1(g: int) -> A: ...
+    @overload
+    def f1(g: str) -> B: ...
+    def f1(g: Any) -> Any: ...
+elif var_true:
+    @overload
+    def f1(g: int) -> A: ...
+    @overload
+    def f1(g: str) -> B: ...
+    def f1(g: Any) -> Any: ...
+else:
+    @overload
+    def f1(g: int) -> A: ...
+    @overload
+    def f1(g: str) -> B: ...
+    def f1(g: Any) -> Any: ...
+reveal_type(f1(42))  # N: Revealed type is "__main__.A"
```

Comment:
What happens in cases like this:

```py
@overload
def f(x: A) -> A: ...
if x:
    @overload
    def f(x: B) -> B: ...
elif y:
    @overload
    def f(x: C) -> C: ...
else:
    @overload
    def f(x: D) -> D: ...
```

Consider all combinations of bool values for `x` and `y`. For example, if `x` is always false but `y` is always true, should we pick up `C -> C` from the elif block? We should have a test case that covers this. If this isn't supported, there should be a useful error message.